src/mem/mem_scheduler.cc

```diff
@@ -0,0 +1,499 @@
+    return needRetry;
+}
+
+PacketPtr
+MemScheduler::CPUSidePort::getBlockedPkt(){
+    return blockedPacket;
+}
+
+void
+MemScheduler::CPUSidePort::recvFunctional(PacketPtr pkt)
+{
+    // Just forward to the memobj.
+    return owner->handleFunctional(pkt);
+}
+
+bool
+MemScheduler::CPUSidePort::recvTimingReq(PacketPtr pkt)
+{
+    // Just forward to the memobj.
+    if (!owner->handleRequest(this, pkt)) {
+        DPRINTF(MemScheduler, "recvTimingReq: handleRequest returned false.\n");
+        needRetry = true;
+        return false;
+    } else {
+        return true;
+    }
+}
+
+void
+MemScheduler::CPUSidePort::recvRespRetry()
+{
+    // We should have a blocked packet if this function is called.
+    assert(blockedPacket != nullptr);
+
+    // Grab the blocked packet.
+    PacketPtr pkt = blockedPacket;
+    DPRINTF(MemScheduler, "recvRespRetry: Changing blockedPacket value, blockedPacket == nullptr: %d\n", blockedPacket == nullptr);
+    blockedPacket = nullptr;
+    DPRINTF(MemScheduler, "recvRespRetry: Changed blockedPacket value, blockedPacket == nullptr: %d\n", blockedPacket == nullptr);
+
+    // Try to resend it. It's possible that it fails again.
+    sendPacket(pkt);
+}
+
+void
+MemScheduler::MemSidePort::sendPacket(PacketPtr pkt)
+{
+    // Note: This flow control is very simple since the memobj is blocking.
+    panic_if(blocked == true, "Should never try to send if blocked MemSide!");
+    // If we can't send the packet across the port, store it for later.
+    if (!sendTimingReq(pkt)) {
+        blockedPacket = pkt;
+        DPRINTF(MemScheduler, "Setting blocked to true on port %s\n", this->name());
+        blocked = true;
+    }
+    DPRINTF(MemScheduler, "Send packet 111\n");
+}
+
+bool
+MemScheduler::MemSidePort::recvTimingResp(PacketPtr pkt)
+{
+    // Just forward to the memobj.
+    return owner->handleResponse(pkt);
+}
+
+bool MemScheduler::MemSidePort::getBlocked()
+{
+    return blocked;
+}
+void
+MemScheduler::MemSidePort::recvReqRetry()
+{
+    // We should have a blocked packet if this function is called.
+    assert(blocked == true && blockedPacket != nullptr);
+    DPRINTF(MemScheduler, "Received ReqRetry\n");
+    // Try to resend it. It's possible that it fails again.
+    blocked = false;
+    sendPacket(blockedPacket);
+    blockedPacket = nullptr;
+
+    owner->wakeUp();
+}
+
+void
+MemScheduler::MemSidePort::recvRangeChange()
+{
+    owner->sendRangeChange();
+}
+
+bool
+MemScheduler::handleRequest(CPUSidePort *port, PacketPtr pkt)
+{
+    panic_if(!(pkt->isRead() || pkt->isWrite()),
+             "Should only see read and writes at memory controller\n");
+
+    uint32_t requestorId = pkt->req->requestorId();
+
+    std::unordered_map<RequestorID, CPUSidePort*>::const_iterator ret = routingTable.find(requestorId);
+    if (ret == routingTable.end())
+        routingTable[requestorId] = port;
+
+    std::unordered_map<RequestorID, bool>::const_iterator rit = readBlocked.find(requestorId);
```

Comment:
Initializing these maps should probably be done in `::startup()`. I believe that you can query the system object (you can add this as a parameter with `parent.Any` as the default proxy) to get the number of requestors in the system.